Feature/webcms 4277 collective.ckeditor wieder vervollstandigen

veritazx opened the pull request:
1) Workover requireJS / mockup / jarn in a more compatible way
2) bring back the inner css for the ckeditor canvas (iframe) by removing the portal_css parts in the view and replacing it with equally specific internal stuff from the styles viewlet that uses the new registry. Could need some improvement.

veritazx committed:
WEBCMS-4277 js-translations need to be Plone4/5 compatible (using mockup or jarn.jsi18n)

veritazx committed:
WEBCMS-4277 remove direct use of portal_css and the dependency to Products.ResourceRegistries

veritazx commented:
I was unsure whether a viewlet lookup via the viewlet manager was too unflexible. 
Since the css links always need to be in the &lt;head&gt;, the Viewlet generating them will always be in the htmlhead viewlet manager. Ergo the way this code works is fine. Please review and merge :)

veritazx committed:
mark the response of getCK_vars as application/x-javascript

veritazx committed:
ensure the alerts element exists when posting alerts into it

veritazx committed:
improve i18n handling

veritazx commented:
Added three small fixes to prevent js errors in different edge cases

pgrunewald commented on src/collective/ckeditor/browser/ckeditorview.py:
```diff
@@ -184,35 +186,25 @@ def getCK_contentsCss(self):
+        # viewlet managers are found by Multi-Adapter lookup
+        adapter = queryMultiAdapter((self.portal, self.request, view), IViewletManager, manager_name, default=None)
+        if adapter:
+            # TODO: can we lookup the viewlet directly? (what if it was transferred to another manager?)
```
There is indeed a way, see: https://docs.plone.org/develop/plone/views/viewlets.html

That way omit the manager registration. The end result is identical, although we need a helper method.

veritazx replied:
I am not completely persuaded yet. I thought, the viewlet-manager-way was faster because the viewlet already got initialised (though i am not a 100% sure it was). With the other way (get registration, instantiate viewlet, use viewlet), the viewlet is newly initialised and loads all the stuff from the registration, that has already been loaded once. Should I still switch to that way?

veritazx replied:
Ah. I didn't realise we are in a completely separate request (for the iframe content). So the viewlet has not been instantiated yet. I'll switch to the method shown in the documentation.

pgrunewald commented on src/collective/ckeditor/browser/ckeditorview.py:
```diff
@@ -184,35 +186,25 @@ def getCK_contentsCss(self):
-        return css_jsList
+        css_jsList = []
+        # retrieve plone.htmlhead.links viewlet manager
+        view = View(self.portal, self.request)
```
No need to register a view, when we have already one (namely ```self```).

veritazx replied:
Changed.

pgrunewald commented on src/collective/ckeditor/browser/statics/ckeditor_plone.js:
```diff
@@ -64,29 +66,43 @@ if ( window.CKEDITOR )
             // Show a friendly compatibility message as soon as the page is loaded,
             // for those browsers that are not compatible with CKEditor.
-            if ( !CKEDITOR.env.isCompatible )
+            if ( CKEDITOR.env.isCompatible )
```
The explamation mark got removed accidently, right? The messages are telling, that something is incompatible, yet the browser IS to be known to be NOT incompatible.

veritazx replied:
Whoops. Indeed. Tested that and forgot.

pgrunewald commented on src/collective/ckeditor/browser/statics/ckeditor_plone.js:
```diff
@@ -64,29 +66,43 @@ if ( window.CKEDITOR )
+                require(['mockup-i18n'], function(I18N) {
                     if (typeof(i18n) === 'undefined') {
-                        var i18n = new I18N();
+                        i18n = new I18N()
```
Missing semicolon

veritazx replied:
Done.

pgrunewald commented on src/collective/ckeditor/browser/statics/ckeditor_plone.js:
```diff
@@ -64,29 +66,43 @@ if ( window.CKEDITOR )
-            require(['mockup-i18n'], function(I18N) {
-                try {
+            /* Setting up jsi18n message factory for ckeditor with jsi18n (from mockup) via requirejs.
+                Use a fallback for plone 4 without require but
```
This sentence abruptly ends...

veritazx replied:
Done.

pgrunewald commented on src/collective/ckeditor/browser/statics/ckeditor_plone.js:
```diff
@@ -64,29 +66,43 @@ if ( window.CKEDITOR )
+                try {
+                    init(jarn.i18n);
+                } catch(e) {
                     console.log('failed to load i18n');
```
Some sort of context information whould be nice, which hints the developper that it happens in collective.ckeditor.

veritazx replied:
Done.

pgrunewald reviewed (changes requested):
The code works as intended, but can be improved nonetheless. Especially we can skip the manager for getting the viewlet. I also have some more generall points.

veritazx committed:
WEBCMS-4277: fix typos, add infos, cleanup

veritazx committed:
WEBCMS-4277 improve styles viewlet handling

- use the viewlet lookup suggestion from the plone viewlet documentation
- eliminate the need to know where the viewlet resides (in which manager)
- add some more comments to the complete method

pgrunewald reviewed (approved):
LGTM
